### ISSUE TITLE
gRPC xDS client stops working after a period of inactivity

### ISSUE THREAD

**Issue by rdesgroppes** [wongnai/xds#11]
We've been using your [xDS Server for gRPC on Kubernetes](https://github.com/wongnai/xds), which suits our needs and is easy to use.
However, we are experiencing problems when the channels switch to idle mode. Disabling idleness helps, but we'd now like to restore the default behavior.

I'm far from being an expert on the matter, that's why I'm kindly requesting your point of view on the issue that was originally filed against the gRPC-Java client: https://github.com/grpc/grpc-java/issues/9632.

According to https://github.com/grpc/grpc-java/issues/9632#issuecomment-1317446356:
> [...] This seems like a go-control-plane bug to me. Once grpc-java sends a request with the removed resources, go-control-plane needs to honor that; it should not need any subsequent ACK/NACK. [...]

Is the `xDS Server for gRPC on Kubernetes` behaving like the aforementioned `go-control-plane`?

**Comment by whs** [comment]
Internally this software use go control plane, so any bugs by them will have to be fixed upstream.

Not sure if related but we do find idle bugs in grpc go, and currently our workaround is to pin grpc at a specific version that do not have revamped idle behavior

**Comment by rdesgroppes** [comment]
> Not sure if related but we do find idle bugs in grpc go, and currently our workaround is to pin grpc at a specific version that do not have revamped idle behavior

Right, idleness is enabled by default since [gRPC-Go 1.59.0](https://github.com/grpc/grpc-go/releases/tag/v1.59.0). We had to create channels with an additional option: `grpc.WithIdleTimeout(0)`.

**Comment by hanikesn** [comment]
We're facing a very similar issue using grpc v1.65.0, where channels who resume from idle mode won't resolve endpoints via xDS correctly.

I'm wondering whether that's something worth reporting upstream.

**Comment by hanikesn** [comment]
Looks like this is the upstream issue to track:
- https://github.com/envoyproxy/go-control-plane/issues/431

**Comment by hanikesn** [comment]
Upstream fix is on the way:
- https://github.com/envoyproxy/go-control-plane/pull/1293
- https://github.com/envoyproxy/go-control-plane/pull/1301

**Comment by hanikesn** [comment]
@whs There's a release including the fix now :
https://github.com/envoyproxy/go-control-plane/releases/tag/xdsmatcher%2Fv0.14.0

**Comment by whs** [comment]
Thanks for the ping. I've saw the release a while ago. However, it is still stuck on the bottom of my TODO list (usually before we cut a release we test it out internally for a few days, so it is not just bump deps and release)

**Comment by whs** [comment]
A bit of an update, I'm planning the change to be done after the end of [a huge government campaign](https://en.moneyandbanking.co.th/2025/209680/) (which may be extended into January 2026) to avoid any disruption during the period.

If you need the change to be done before then, please feel free to submit a PR but we might not be able to commit any testing to be done before the campaign.

**Comment by whs** [comment]
v1.4.0-pre1 is tagged in the container registry. We're running smoke testing internally before merging the PR and cutting a release (probably in Jan 2026).

**Comment by whs** [comment]
I totally forgot about this until another internal team run into problem.

v1.5.0-pre1 is out. We shipped it to production yesterday so it probably be safe for you to use it too. If I don't forget and we don't have issues then I'll tag v1.5.0 proper.

**Comment by rdesgroppes** [comment]
Thanks, @whs!

**Comment by whs** [comment]
SRE reports that the memory usage of our latest XDS deployment upgrade has grown from ~300MB to ~900MB. If you can pay the memory cost, then it should be usable (we used it in prod for a long while now)

Claude think the root cause is https://github.com/envoyproxy/go-control-plane/pull/1324
